### ISSUE TITLE
Implement coherent JavaScript cache key generator to support any arguments

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#406]
The current mechanism for generating cache keys for data requests to the REST API relies on whitelisting specific arguments only which limits which kinds of requests can be made (while still working correctly). This limitation has led to some datapoints being almost identical and only existing for the sake of one minor tweak, because passing this tweak via an argument hasn't been possible.

A secondary requirement of this enhancement is that it should be possible to easily invalidate the cache for any of the following hierarchical levels:
* `dataObject` / `type`
* `identifier`
* `datapoint`

In other words, something like `deleteCache( 'modules', 'analytics', 'get-accounts' )` should wipe all cache keys for that datapoint, while something like `deleteCache( 'modules', 'search-console' )` should wipe all cache keys for all datapoints of the Search Console module.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation Brief

* 

## Changelog entry

*

**Comment by aaemnnosttv** [comment]
@felixarntz - a few questions for you on this for clarification, as well as a few suggestions as the end. There is a lot of information to review here but the AC is also very much written in terms of IB. Can you review this as well and make sure all implementation details are in the brief section? 

---

> The getCache() function parameters should be simplified to `getCache( key, maxAge = 0 )`.

This implies that the function would return never return anything when just passing the key as all cache entries would almost certainly have an `age > 0` when this is called. This invalidates the "optional" quality of the maxAge parameter. I would think maxAge should default to a value which would not restrict the age: e.g.  `-1` (my preference), or if there is such a thing, a default maxAge value (1 day, etc).

> If data is passed and the object is not empty, sort the object properties, JSON-encode the object, and pass the output through the md5() function. Append the result to the cache key, separated with another ::.

This seems like an unnecessary condition. Why not hash the object the same regardless of its contents?

> Update all JS code relying on combinedGet() (most areas do that via withData()) to pass all extra request data in a single data object.

This is a bit unclear. Does this have to do with the `paramsToMigrate` and multiple formats in use? Does this imply the `googlesitekit.dateRange` filter would be removed by requiring this parameter from requests?

---

Suggestion: Refactor `noCache` to `withCache` (invert logic) as it makes negations easier to read by avoiding double negatives: e.g. `! noCache` would be `useCache`, `noCache` would be `! useCache` 👌 

Suggestion: Refactor cache option from a single parameter to an object of options. This is much better for readability (and closer to [WP standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#self-explanatory-flag-values-for-function-arguments) (no example in JS standards)) by avoiding booleans as parameters. It also makes the signature more flexible to adding options in the future.

**Comment by felixarntz** [comment]
@aaemnnosttv Regarding your suggestions:

> Suggestion: Refactor `noCache` to `withCache`

Makes sense, however I excluded changing this from here, as I intend to remove the parameter entirely in the new API. There will be a `useCache()` function to toggle this. The cache should always be active and only be disabled in very specific cases, so having this separately gets rid of this "cluttering" function parameter.

> Refactor cache option from a single parameter to an object of options.

Can you clarify what you mean? Are you talking about the values stored in cache?

**Comment by aaemnnosttv** [comment]
> I should have clarified, the 0 would basically be like undefined. The parameter would only be used if it contained a positive number.

Ok, that makes sense 👍 

> My idea was that an empty `data` object would basically be the same as if no data object was provided at all, so they should share the same key.

Agreed! According to the outlined use of `deleteCacheGroup`, `getCacheKey` needs to support returning a key without the hash (since groups don't use `data` in the key). The idea is that all cache entries should have a hash for consistency, but it should still be possible to generate a cache key without it. `data` could default to an empty object in the method signature, and `getCacheKey` could explicitly pass `null`/`false` to disable the object hashing part of the resulting key.  Just an idea. It could also just conditionally add the hash as you had proposed. I think it's generally safer to err on the side of consistency, all things being equal.

> There will be a useCache() function to toggle this

As in globally toggling cache on/off? If so, I would think that may be a bit heavy-handed. Would it not be more appropriate to toggle it per-request as it is now?

> Can you clarify what you mean? Are you talking about the values stored in cache?

Sure, instead of this:

https://github.com/google/site-kit-wp/blob/673e6dfb02c89cfd45f462d9135a53decbf235da/assets/js/components/data.js#L445

Using an options object instead

```js
function get( dataObject, identifier, datapoint, queryArgs = {}, options = { nocache: true }  )
```

Then in use, the the boolean is self-documenting
```js
return get( 'modules', 'analytics', 'get-accounts', {}, { nocache: false }  )
```

**Comment by felixarntz** [comment]
> `data` could default to an empty object in the method signature, and `getCacheKey` could explicitly pass `null`/`false` to disable the object hashing part of the resulting key. Just an idea. It could also just conditionally add the hash as you had proposed. I think it's generally safer to err on the side of consistency, all things being equal.

You're actually reminding myself now, why I decided to handle an empty object the same as no data: Precisely because I wanted to be able to easily create a object-hash-less key. I'd prefer to leave the logic as proposed, so that we don't need an extra `disableDataHash` parameter. With the proposed logic, the function accounts for both needs out of the box.

> > There will be a useCache() function to toggle this
> 
> As in globally toggling cache on/off? If so, I would think that may be a bit heavy-handed. Would it not be more appropriate to toggle it per-request as it is now?

Let's discuss this separately. My perfect scenario would actually be that you can't disable cache lookup at all, because the caching exists for a reason. :) With the ability to delete cache groups, we might actually be able to pull this off while still allowing the existing edge-cases to work (by using that new `deleteCacheGroup`).

> Sure, instead of this:
> 
> https://github.com/google/site-kit-wp/blob/673e6dfb02c89cfd45f462d9135a53decbf235da/assets/js/components/data.js#L445
> 
> Using an options object instead
> 
> ```js
> function get( dataObject, identifier, datapoint, queryArgs = {}, options = { nocache: true }  )
> ```
> 
> Then in use, the the boolean is self-documenting
> 
> ```js
> return get( 'modules', 'analytics', 'get-accounts', {}, { nocache: false }  )
> ```

I like that idea, if we keep the `nocache` functionality. I'd prefer to eliminate it entirely as mentioned, and then there wouldn't be a need for any other parameter. _If_ we ever want to allow specific "flags" for these functions, then I absolutely agree it should happen in an options object. Let's revisit this when we get to designing the new API `googlesitekit.api`.

**Comment by aaemnnosttv** [comment]
Sounds good @felixarntz, that's all I had. Implementation brief is good 👍

**Comment by adamsilverstein** [comment]
I did some testing on this looks really good! In particular the caching behaviour is as expected and pages loaded correctly, date selectors worked no errors anywhere.

* Initial load on Dashboard resulted in two batch requests (one is a low priority request for page speed insights) as expected
* secondary loads came from cache (no rest request)
* verified cache keys consistent.
* module pages loaded with one request, and none on reloads.

Some questions/issues:

* Noticed a few tiny doc block typos, PR: https://github.com/google/site-kit-wp/pull/504
* The acceptance criteria mentioned adding a `deleteCacheGroup` function which I don't see. Was that forgotten, dropped, or planned for a future PR?

**Comment by aaemnnosttv** [comment]
> The acceptance criteria mentioned adding a deleteCacheGroup function which I don't see. Was that forgotten, dropped, or planned for a future PR?

I believe it was implemented as `invalidateCacheGroup()` instead
https://github.com/google/site-kit-wp/blob/5ee8d08e0775f5411caeedf126646c4bc923c114/assets/js/components/data.js#L353-L360

I'm not sure why the name change, I think `deleteCacheGroup` may actually be more fitting in this case since the current implementation actually deletes the existing values. To me, invalidating is leaving the values in the cache, and altering something like a cache version which is baked into the cache key that causes future generations of a cache key with the same args to generate a different one. I think this is more for if situations where deleting would be too expensive or if it would be not possible to identify the group by the key alone (e.g. if the whole key were a hash).

It's not a big deal I think, but I'm also curious about the change.

**Comment by adamsilverstein** [comment]
> I believe it was implemented as invalidateCacheGroup() instead

Aha, thanks for clarifying - that seems fine.

I did notice one other regression in develop related to https://github.com/google/site-kit-wp/pull/486

I am unable to complete AdSense setup, the setup flow hangs here:

![image](https://user-images.githubusercontent.com/2676022/64463626-e756c600-d0c1-11e9-9bbf-b34dbbb05498.png)

According to `git bisect` this problem starts after https://github.com/google/site-kit-wp/commit/bb3702c50d06078e7f7592cd2328cdc2a54f0864 cc: @aaemnnosttv

**Comment by felixarntz** [comment]
Thanks for all the testing @adamsilverstein - can either you or @aaemnnosttv open a PR (you can associate it with #486) to fix the AdSense issue? Should be quick, just check `existingTag && existingTag.length` (or maybe we can even remove this since the non-found value is falsey now anyway?).

**Comment by aaemnnosttv** [comment]
No problem! Will do, although I think checking the length shouldn’t be necessary as a zero length string is also falsey.

**Comment by adamsilverstein** [comment]
I'll work on a PR to address the AdSense setup issue.

**Comment by adamsilverstein** [comment]
https://github.com/google/site-kit-wp/pull/507 resolves the issue noted in my comment above https://github.com/google/site-kit-wp/issues/406#issuecomment-529034240

**Comment by adamsilverstein** [comment]
@felixarntz This passes my QA now, once #507 is merged. Unless you see any outstanding issues, we can move this to "Approval" once that is merged.
